### PR TITLE
Support ReadTheDocs

### PR DESCRIPTION
**Pull request by jdemeyer**
http://fpylll.readthedocs.io/en/latest/

### CODE REVIEW

**Comment by malb** [comment]
Ta! So, I guess, once merged we'd have to make readthedocs point to this repository?

**Comment by jdemeyer** [comment]
> Ta! So, I guess, once merged we'd have to make readthedocs point to this repository?

Yes. I still have to figure out how to do that :-)

**Comment by malb** [comment]
Why did you enable binding in `setup.py`?

File: setup.py [hunk 1]
```diff
@@ -28,6 +37,11 @@
          "library_dirs": library_dirs,
          "libraries": ["gmp"]}
 
+if "READTHEDOCS" in os.environ:
+    # ReadTheDocs uses fplll from Conda, which was built with the old
+    # C++ ABI.
+    fplll["extra_compile_args"].append("-D_GLIBCXX_USE_CXX11_ABI=0")
```
Comment (isuruf) [inline setup.py]: FYI, there's a `gcc` package in conda that installs `4.8.5`.

**Comment by jdemeyer** [comment]
> Why did you enable binding in setup.py?

Interesting question!

The `binding` directive in Cython changes functions defined in Cython from "built-in functions" to a Cython-specific type called `cyfunction`. These cyfunctions behave much more like plain Python functions. In particular, they support inspection (argspec and so on), which is needed for proper Sphinx autodoc support.

As you may know, SageMath has a lot of custom code to deal with inspection of Cython code. Most of that could be replaced by compiling Cython code with `binding=True`.

There is one important caveat: the Python `inspect` module does not consider cyfunctions as functions. That is why I pointed `suggestions.txt` to use my fork of Sphinx which basically changes
```
inspect.isfunction(foo)
```
to
```
"__code__" in dir(foo)
```
which is a more duck-typing way to check for functions and which works for cyfunctions. See https://github.com/sphinx-doc/sphinx/pull/3623

The original reason for the name "`binding`" is because cyfunctions support the descriptor protocol (`__get__`), so you can do stuff like
```
def somefunc(self):
    ...

class X:
    method = somefunc
```
but `binding` is more than just that.

**Comment by jdemeyer** [comment]
> FYI, there's a gcc package in conda that installs 4.8.5.

Does that include a C and C++ compiler (i.e. `gcc` and `g++`)?

**Comment by isuruf** [comment]
> Does that include a C and C++ compiler (i.e. gcc and g++)?

Yes, `gcc, g++, gfortran and gcj`

**Comment by malb** [comment]
Okay, figured out how to deal with the doctest failures. Let me know if you want to augment the PR with the gcc tweak first or if it's cool to merge (I'll fix up test failures once merged)

**Comment by jdemeyer** [comment]
I'll try `gcc` in Conda.

**Comment by jdemeyer** [comment]
> Yes, gcc, g++, gfortran and gcj

It didn't work: https://readthedocs.org/projects/fpylll/builds/5291372/

**Comment by jdemeyer** [comment]
I just reverted to the previous version (without `gcc` from conda)

**Comment by isuruf** [comment]
Sorry about that. I haven't seen that error before. There's a long thread here, but no solution apparently. https://github.com/ContinuumIO/anaconda-issues/issues/1392

**Comment by jdemeyer** [comment]
You broke the build again in

https://github.com/fplll/fpylll/commit/87751d0e33c73d6f3e528d5fe64ba6af697b3550

**Comment by malb** [comment]
Should be fixed now. Thanks! Btw. is `embedsignature` still needed? You dropped it, I kept it.

**Comment by malb** [comment]
Done. Can you make me an owner of the ReadTheDocs project

**Comment by jdemeyer** [comment]
> Done. Can you make me an owner of the ReadTheDocs project

What is your ReadTheDocs username?

**Comment by malb** [comment]
martinralbrecht

**Comment by jdemeyer** [comment]
On 2017-04-16 16:09, Martin Albrecht wrote:
> martinralbrecht

I just added you as maintainer of the fpylll docs.

**Comment by malb** [comment]
Ta

**Pull request merged**